### PR TITLE
Updating the media-streaming benchmark

### PR DESCRIPTION
**Pull request by aansaarii**
In this PR, I've added the ability to fix the number of Nginx workers after the Nginx installation. The default number of workers might not be sufficient to handle the clients' load. Using this PR, the server can set the appropriate number of workers to smaller or larger numbers based on the server's capabilities.

### CODE REVIEW

**Review by Xusine1131 — CHANGES_REQUESTED** [review]
I think we should also update the document so that the potential users know they can set Nginx works.

**Comment by aansaarii** [comment]
A new commit is added to this PR: "fixed port allocation of httperf"
The problem is that httperf creates a [map](https://github.com/parsa-epfl/cloudsuite/blob/main/benchmarks/media-streaming/client/files/videoperf/core.c#L80) to keep track of the ports that it allocates. Httperf [lookups](https://github.com/parsa-epfl/cloudsuite/blob/main/benchmarks/media-streaming/client/files/videoperf/core.c#L1161) this map to find a free port to assign it to a connection. However, another process on the machine might already allocate the found port. Consequently, the binding phase would fail and httperf searches for another free port (check [here](https://github.com/parsa-epfl/cloudsuite/blob/main/benchmarks/media-streaming/client/files/videoperf/core.c#L1172) and [here](https://github.com/parsa-epfl/cloudsuite/blob/main/benchmarks/media-streaming/client/files/videoperf/core.c#L1159)). The point is that the found port is already [flagged](https://github.com/parsa-epfl/cloudsuite/blob/main/benchmarks/media-streaming/client/files/videoperf/core.c#L315) as not free during the lookup phase and if httperf cannot use this port, httperf will never have a chance to set the port as free in the future. As a result, httperf will run out of ports gradually while the ports are actually available on the machine because the processes that were using those ports have released the ports. I noticed this problem when I tried to run the benchmark with multiple httperf instances at the same time with high loads for around 10 minutes. The new commit updates the map as a port is allocated when the binding phase successfully binds the free port to the connection.

File: benchmarks/media-streaming/client/files/docker-entrypoint.sh
Comment (Hnefi) [inline benchmarks/media-streaming/client/files/docker-entrypoint.sh:7]: If we are going to make this change to the benchmark, we should update all of the documentation to explain what these parameters are, and how the user can adjust them.

File: benchmarks/media-streaming/client/files/run/peak_hunter/launch_remote.sh
Comment (Hnefi) [inline benchmarks/media-streaming/client/files/run/peak_hunter/launch_remote.sh:28]: Do we really want to make verbose the default option? I would say no.

File: benchmarks/media-streaming/client/files/videoperf/core.c
Comment (Hnefi) [inline benchmarks/media-streaming/client/files/videoperf/core.c:315]: I can't say I understand this change, it's not obvious to me what is the problem in the code. Why does your change to _not_ set bits in the port map solve it? It seems like all that this change does is "shoot the messenger" for lack of a better term, because the function `get_port()` can still return a valid answer without setting the port map structure.

File: benchmarks/media-streaming/client/files/videoperf/core.c
Comment (Hnefi) [inline benchmarks/media-streaming/client/files/videoperf/core.c:1179]: See above comment.

File: benchmarks/media-streaming/client/files/videoperf/httperf.c
Comment (Hnefi) [inline benchmarks/media-streaming/client/files/videoperf/httperf.c:82]: See above comment - why should we make verbose printing the default option?

File: benchmarks/media-streaming/client/files/videoperf/stat/basic.c
Comment (Hnefi) [inline benchmarks/media-streaming/client/files/videoperf/stat/basic.c:220]: These two calculations are the same.

**Review by Hnefi — CHANGES_REQUESTED** [review]
Hi Ali,
Thanks for your work so far on this. I have added a bunch of comments into the individual changes, which we should fix before we merge anything. You can address them as you get time.

Also, we should change the name of this PR to something correct, because it currently contains many more changes than just setting the nginx worker count.

File: benchmarks/media-streaming/client/files/docker-entrypoint.sh
Comment (aansaarii) [reply benchmarks/media-streaming/client/files/docker-entrypoint.sh:7]: The doc file provides information about these input values.

File: benchmarks/media-streaming/client/files/videoperf/core.c
Comment (aansaarii) [reply benchmarks/media-streaming/client/files/videoperf/core.c:315]: Did not the comment that I wrote above clarify the need for this change? If httperf sets the bit here and later on, it notices that it cannot bind the connection to the found port, it will never call `core_close` that calls `put_port` that resets the port bit to make it available for the future connections. The `port_get` function has to only suggest a port from its free list and setting the bit has to be done after httperf detects no other process on the machine uses that port to safely allocate it (which I did it [here](https://github.com/parsa-epfl/cloudsuite/blob/fix_streaming_server/benchmarks/media-streaming/client/files/videoperf/core.c#L1179)) and then, release the port by resetting the bit whenever the connection is closed.

File: benchmarks/media-streaming/client/files/videoperf/httperf.c
Comment (aansaarii) [reply benchmarks/media-streaming/client/files/videoperf/httperf.c:82]: Verbose is removed.

**Comment by aansaarii** [comment]
A new commit is added to update the TLS version from 1.2 to 1.3.

File: benchmarks/media-streaming/client/files/docker-entrypoint.sh
Comment (Hnefi) [reply benchmarks/media-streaming/client/files/docker-entrypoint.sh:7]: Good. However, their explanation does not really match the graphs we have posted on the wiki. We have the rate on the x axis, but this only dictates the "ramp up" rate on the server. The total constant load is the number of sessions, which is what should be on the x-axis?

File: benchmarks/media-streaming/client/files/videoperf/stat/basic.c
Comment (Hnefi) [reply benchmarks/media-streaming/client/files/videoperf/stat/basic.c:220]: Why do we want to change the benchmark from MB to MiB? It's not measuring binary-addressable things like caches.

File: benchmarks/media-streaming/server/Dockerfile
Comment (Hnefi) [inline benchmarks/media-streaming/server/Dockerfile:9]: Why is this changed? Is it a typo?

Comment (aansaarii) [reply benchmarks/media-streaming/server/Dockerfile]: Yes. Sorry. I pushed another commit to fix it.

File: benchmarks/media-streaming/client/files/videoperf/core.c
Comment (Hnefi) [reply benchmarks/media-streaming/client/files/videoperf/core.c:315]: After discussion, this change only marks ports as allocated when _this process_ successfully binds to them.

The implications are that other processes may have bound to ports that this process believes are available, and therefore more syscalls to open ports may fail. This may or may not impact performance, which needs to be tested.

File: docs/benchmarks/media-streaming.md
Comment (Hnefi) [inline docs/benchmarks/media-streaming.md:37]: We need to improve this explanation:
- Add a general description of dataset access log session generator. How it works, and how the clients use it.
- Also be clear on the fact these session are not the dynamic number that the client uses, these are just all videos that can possibly be accessed. The client loops over this "session list" to access certain videos.

File: docs/benchmarks/media-streaming.md
Comment (Hnefi) [inline docs/benchmarks/media-streaming.md:66]: We need to be a bit more clear here:
- Number of individual sessions per client process, or number of sessions total?
- Forward pointer from dataset part to this part, explaining how the clients process the log.
- We should specify how the steady state works (does the client keep generating sessions or not?) and when the ramp down happens.

**Review by Hnefi — CHANGES_REQUESTED** [review]
We can merge this after the following:
- Let's address the comments in my review for documentation clarity.
- We should fix the tests.

**Review by Xusine1131 — DISMISSED** [review]

**Review by Hnefi — APPROVED** [review]
I have looked at as many of the changes as is possible, and approved the merging with two follow-ups (as discussed):

1. It would be good to divide the documentation into a README file which describes simply how to run the benchmark, and a more detailed document (perhaps a PDF) with the architecture of the load generator, how to run an experiment, and how to set various parameters to create certain scenarios. This would give more structure to each file (i.e., the README which is visible by default just explains the architecture of the benchmark and how to run it, and then more details are available for those who are curious to tune it).
2. We should attempt in future to set all editors so that they do not autoformat/change spacing of code. For example, it is not possible to review the changes to files such as `core.c` or `videoperf.c` because the changelist currently is polluted with simply re-spacing files or converting spaces back and forth to tabs.

**Pull request merged**
